src/frontmatter.md

```diff
@@ -0,0 +1,42 @@
+r[frontmatter]
+# Frontmatter
+
+r[frontmatter.syntax]
+```grammar,lexer
+FRONTMATTER ->
+      FRONTMATTER_FENCE INFOSTRING? LF
+      (FRONTMATTER_LINE LF )*
+      FRONTMATTER_FENCE[^matched-fence] LF
+
+FRONTMATTER_FENCE -> `---` `-`*
+
+INFOSTRING -> XID_Start ( XID_Continue | `.` )*
+
+FRONTMATTER_LINE -> (~INVALID_FRONTMATTER_LINE_START (~INVALID_FRONTMATTER_LINE_CONTINUE)*)?
+
+INVALID_FRONTMATTER_LINE_START -> (FRONTMATTER_FENCE[^escaped-fence] | LF)
+
+INVALID_FRONTMATTER_LINE_CONTINUE -> LF
```

Comment:
Was unsure whether to create a new rule for non-newline whitespace and specify that in every location or not.